### PR TITLE
Download all binaries instead of building them locally for compat version tests

### PR DESCRIPTION
**Pull request by michaelasp**
This pr downloads all required docker and test binaries and uses those to run the tests rather than building them locally. Running these tests locally showed a major performance improvement from this. 

The two major changes are downloading all the control plane docker images rather than building them from source and also downloading the test binaries instead of building them to run conformance tests.

### CODE REVIEW

File: experiment/compatibility-versions/kind-upgrade.sh [hunk 1]
```diff
@@ -66,9 +67,9 @@ parse_args()
 
 parse_args $*
 
-build_docker(){
-  build/run.sh make all WHAT="cmd/kubectl cmd/kubelet" 1> /dev/null
-  make quick-release-images 1> /dev/null
+download_docker(){
```
Comment (BenTheElder) [inline experiment/compatibility-versions/kind-upgrade.sh]: this name is misleading (the old one is too, it's a carry over from when we had a bazel build mode, but this is even more misleading)

maybe "download_images"

File: experiment/compatibility-versions/common.sh [hunk 2]
```diff
@@ -235,6 +235,30 @@ build_prev_version_bins() {
...
+    return 1
+  fi
+  tar -xvf kubernetes-test-$(go env GOOS)-$(go env GOARCH).tar.gz 
+  mkdir -p _output/bin
+  mv kubernetes/test/bin/* _output/bin
+  return 0
+}
+
+download_current_version_bins() {
+  wget https://dl.k8s.io/ci/$(curl -Ls https://dl.k8s.io/ci/latest.txt)/kubernetes-test-$(go env GOOS)-$(go env GOARCH).tar.gz
```
Comment (BenTheElder) [inline experiment/compatibility-versions/common.sh]: maybe something like this:
```suggestion
  wget 'https://dl.k8s.io/ci/'"${1}"/'kubernetes-test-'"$(go env GOOS)-$(go env GOARCH)"'.tar.gz'
```

and pass in the version (see below)

File: experiment/compatibility-versions/common.sh [hunk 3]
```diff
@@ -235,6 +235,30 @@ build_prev_version_bins() {
   echo "Finished building e2e.test binary from ${PREV_RELEASE_BRANCH}."
 }
 
+download_prev_version_bins() {
+  local version=$1
+  wget https://dl.k8s.io/release/$(curl -Ls https://dl.k8s.io/release/stable-${version}.txt)/kubernetes-test-$(go env GOOS)-$(go env GOARCH).tar.gz
```
Comment (BenTheElder) [inline experiment/compatibility-versions/common.sh]: TODO: this script as-written will only work on a linux host, but it should work on mac too.

For that you need to split binaries being downloaded for kind (GOOS should always be linux) versus for the host (like the test binaries)

File: experiment/compatibility-versions/kind-upgrade.sh [hunk 4]
```diff
@@ -32,8 +32,9 @@ CONTROL_PLANE_COMPONENTS="kube-apiserver kube-controller-manager kube-scheduler"
 KUBE_ROOT="."
 # KUBE_ROOT="$(go env GOPATH)/src/k8s.io/kubernetes"
 source "${KUBE_ROOT}/hack/lib/version.sh"
-kube::version::get_version_vars
-DOCKER_TAG=${KUBE_GIT_VERSION/+/_}
+LATEST_IMAGE=$(curl -Ls https://dl.k8s.io/ci/latest.txt)
```
Comment (BenTheElder) [inline experiment/compatibility-versions/kind-upgrade.sh]: we should make sure to use this value when downloading the current version, otherwise we're going to wind up with a race where we record one version and test another.
```suggestion
LATEST_VERSION=$(curl -Ls https://dl.k8s.io/ci/latest.txt)
```

**Comment by BenTheElder** [comment]
Thanks, sorry for the delay (PRR crunch...), some comments above ...

File: experiment/compatibility-versions/kind-upgrade.sh [hunk 5]
```diff
@@ -66,9 +67,9 @@ parse_args()
 
 parse_args $*
 
-build_docker(){
-  build/run.sh make all WHAT="cmd/kubectl cmd/kubelet" 1> /dev/null
-  make quick-release-images 1> /dev/null
+download_docker(){
+  curl -L 'https://dl.k8s.io/ci/'${LATEST_IMAGE}'/kubernetes-server-linux-amd64.tar.gz' > kubernetes-server-linux-amd64.tar.gz
+  tar -xvf kubernetes-server-linux-amd64.tar.gz
```
Comment (BenTheElder) [inline experiment/compatibility-versions/kind-upgrade.sh:70]: You probably want to stage files in a tempdir instead of in the current working directory, it can make the repo dirty (also, they might not get cleaned up ...)

This server tarball can also be used to build a node image, so we don't need to checkout and build k/k at all.

`kind build node-image --type file kubernetes-server-linux-amd64.tar.gz`
https://kind.sigs.k8s.io/docs/user/quick-start/#building-images

File: experiment/compatibility-versions/kind-upgrade.sh (hunk 1 above)
Comment (michaelasp) [reply experiment/compatibility-versions/kind-upgrade.sh]: Done!

File: experiment/compatibility-versions/common.sh (hunk 2 above)
Comment (michaelasp) [reply experiment/compatibility-versions/common.sh]: Makes sense, this can make for some nasty edge cases. I made it so only the create_cluster step obtains the latest_version. I think that's the best place to put it in common.sh since we need to create a cluster in every test and we have access to artifacts at that point.

File: experiment/compatibility-versions/common.sh (hunk 3 above)
Comment (michaelasp) [reply experiment/compatibility-versions/common.sh]: Added a todo :)

File: experiment/compatibility-versions/kind-upgrade.sh (hunk 4 above)
Comment (michaelasp) [reply experiment/compatibility-versions/kind-upgrade.sh]: Ack, made it so only the common script modifies it in the create cluster step.

File: experiment/compatibility-versions/kind-upgrade.sh (hunk 5 above)
Comment (michaelasp) [reply experiment/compatibility-versions/kind-upgrade.sh:70]: Done! I don't think we build k/k at all here, the create cluster step uses a precompiled image. I did however remove the cloning of the k8s repo, this is no longer necessary since we pull the precompiled binary.

File: experiment/compatibility-versions/e2e-skip-version-testing.sh [hunk 6]
```diff
@@ -66,7 +65,9 @@ run_skip_version_tests() {
   # Test removal of emulated version entirely.
   export PREV_RELEASE_BRANCH="release-${EMULATED_VERSION}"
   delete_emulation_version || ret=$?
-  build_prev_version_bins || ret=$?
+
+  LATEST_VERSION="${LATEST_VERSION:-}"
```
Comment (BenTheElder) [inline experiment/compatibility-versions/e2e-skip-version-testing.sh]: Remind me how this is being set?

I see two likely versions, both of which would need addressing:
1. We use the commit we just cloned, which may not actually have a CI build to download from yet.
2. We fall back use the latest version CI has built, but then we record in testgrid the version we just cloned.

If we're going to use the latest CI builds, then we need to add a little patch to record the version to the CI metadata (see the DRA PR), and stop cloning k/k. This is a good solution, it will be faster / cheaper and match the GCE jobs which we run a LOT more of currently.

Alternatively we could keep building from the clone, but then we can't rely on the matching latest binaries existing yet, the build job has some non-zero latency after a commit merges, if this job is triggered before the build has run yet ...

Comment (michaelasp) [reply experiment/compatibility-versions/e2e-skip-version-testing.sh]: This is being set by the common script here https://github.com/kubernetes/test-infra/pull/34930/files#diff-83be0e6d209c4519dcc804ecade1f0be20d0d0fc59e701ba43118bdf8f794131R107-R108

This should run prior to any other test since we source it but in order to pass shellcheck I also have this, maybe we should error if we don't set it? This line in the common script also adds the version to the metadata like the DRA PR.

Comment (BenTheElder) [reply experiment/compatibility-versions/e2e-skip-version-testing.sh]: Yeah, shellcheck is also happy if you reference as `${LATEST_VERSION:?}` to guard against it being unset instead, which will catch the bug if we fail to set it, but doesn't require defaulting to empty.

BTW, if I miss following up on something, feel free to send me a message vs the github request review, github notifications have somewhat poorer signal:noise ratio than DMs 😅

Comment (BenTheElder) [reply experiment/compatibility-versions/e2e-skip-version-testing.sh]: I think we're also still missing recording the version tested? I forget exactly how we need to write that out, but basically $ARTIFACTS/metadata.json, there should be an example in the DRA scripts.

Comment (michaelasp) [reply experiment/compatibility-versions/e2e-skip-version-testing.sh]: Updated and squashed since it's looking close to being ready!

**Comment by michaelasp** [comment]
/cc @siyuanfoundation 

Rebased and should be good for a round of review.

File: experiment/compatibility-versions/common.sh [hunk 7]
```diff
@@ -236,6 +238,32 @@ build_test_bins() {
...
+  mkdir -p _output/bin
+  mv kubernetes/test/bin/* _output/bin
+  export PATH="${PWD}/_output/bin:$PATH"
+  return 0
+}
+
+download_current_version_bins() {
+  wget 'https://dl.k8s.io/ci/'"${1}"/'kubernetes-test-'"$(go env GOOS)-$(go env GOARCH)"'.tar.gz'
+  if [ $? -ne 0 ]; then
+    echo "failed to download previous version binaries"
```
Comment (siyuanfoundation) [inline experiment/compatibility-versions/common.sh]:
```suggestion
    echo "failed to download current version binaries"
```

File: experiment/compatibility-versions/common.sh [hunk 8]
```diff
@@ -236,6 +238,32 @@ build_test_bins() {
   echo "Finished building e2e.test binary from ${release_branch}."
 }
 
+# TODO: Support Mac as a platform along with Linux https://github.com/kubernetes/test-infra/pull/34930#discussion_r2138760351
+download_release_version_bins() {
+  local version=$1
+  wget https://dl.k8s.io/release/$(curl -Ls https://dl.k8s.io/release/stable-${version}.txt)/kubernetes-test-$(go env GOOS)-$(go env GOARCH).tar.gz
+  if [ $? -ne 0 ]; then
+    echo "failed to download previous version binaries"
```
Comment (siyuanfoundation) [inline experiment/compatibility-versions/common.sh]:
```suggestion
    echo "failed to download previous version ${version} binaries"
```

**Comment by siyuanfoundation** [comment]
2 nit comments. 
/lgtm

File: experiment/compatibility-versions/common.sh (hunk 7 above)
Comment (michaelasp) [reply experiment/compatibility-versions/common.sh]: Done!

File: experiment/compatibility-versions/common.sh (hunk 8 above)
Comment (michaelasp) [reply experiment/compatibility-versions/common.sh]: Good idea, added

File: experiment/compatibility-versions/common.sh (hunk 3 above)
Comment (BenTheElder) [reply experiment/compatibility-versions/common.sh]: I think this is actually done now? the test binaries run on the host currently, and we're using GOOS now

we just need to make sure the server / node binaries use linux always, which I think we are already

**Comment by michaelasp** [comment]
Testing locally to make sure nothing broke, will unhold when done.

**Comment by michaelasp** [comment]
Oops, incorrect rebase added a bunch of tags :(

**Review by BenTheElder — COMMENTED** [review]
/lgtm
/approve

**Comment by michaelasp** [comment]
/unhold 
Now that it's Monday lets track how this goes. Will revert/fix asap if something happens since it affects all emulation version tests.

**Comment by Jefftree** [comment]
@michaelasp @BenTheElder: Per the logs, seems like ginkgo is not found https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-e2e-kind-compatibility-versions-n-minus-2/1998591874012549120. Do we need to keep `ginkgo-e2e.sh`?

**Comment by michaelasp** [comment]
> @michaelasp @BenTheElder: Per the logs, seems like ginkgo is not found https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-e2e-kind-compatibility-versions-n-minus-2/1998591874012549120. Do we need to keep `ginkgo-e2e.sh`?

Huh, I thought I fixed this, it should come as a binary in the CI folder. Let me take a look, we can revert if there's no quick fix.

**Comment by michaelasp** [comment]
Hmm... this should be an easy fix, but am I correct in seeing we don't have wget installed. @BenTheElder should we be using `curl` here?

```
+ wget https://dl.k8s.io/ci/v1.33.7-1+fc38c49c0fa93c/kubernetes-test-linux-amd64.tar.gz
/home/prow/go/src/k8s.io/kubernetes/../test-infra/experiment/compatibility-versions/common.sh: line 216: wget: command not found
+ '[' 127 -ne 0 ']'
+ echo 'failed to download previous version 1.33 binaries'
failed to download previous version 1.33 binaries
```
